File: Baekjoon/BFS/Baekjoon11724.py
# www.acmicpc.net/problem/11724
# 백준 11724번 문제: 요소의 개수

import sys
sys.setrecursionlimit(10000)

def dfs(start, visited):
  visited[start] = True
  for i in graph[start]:
    if not visited[i]:
      dfs(i, visited)

n, m = map(int, sys.stdin.readline().split())
graph = [[] for _ in range(n+1)]
visited = [False] * (n+1)
count = 0

for _ in range(m):
  u, v = map(int, sys.stdin.readline().split())
  graph[u].append(v)
  graph[v].append(u)

for i in range(1, n+1):
  if not visited[i]:
    dfs(i, visited)
    count += 1

print(count)

File: Baekjoon/BFS/Baekjoon11724.swift
//
//  Baekjoon11724.swift
//  
//
//  Created by 임현규 on 2024/03/15.
//

import Foundation

var input = readLine()!.split(separator: " ").map { Int($0)! }
let n = input[0], m = input[1]
var graph: [[Int]] = Array(repeating: [], count: n + 1)

for _ in 0..<m {
    input = readLine()!.split(separator: " ").map { Int($0)! }
    let u = input[0] - 1, v = input[1] - 1
    graph[u].append(v)
    graph[v].append(u)
}

func solution(_ graph: [[Int]], _ n: Int) -> Int {
    var result = 0
    var visited = Array(repeating: false, count: n)
    
    func bfs(_ start: Int) {
        var count = 1
        var queue = [start]
        visited[start] = true
        
        while !queue.isEmpty {
            let ns = queue.removeLast()
            
            for i in graph[ns] {
                if !visited[i] {
                    queue.append(i)
                    visited[i] = true
                }
            }
        }
    }
    
    for i in 0..<n {
        if !visited[i] {
            bfs(i)
            result += 1
        }
    }
    
    return result
}

print(solution(graph, n))
